spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,205 @@
+  private static final String UPDATE_AFTER = ChangelogOperation.UPDATE_AFTER.name();
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> identifierFieldIdx;
+
+  private Row cachedRow = null;
+
+  private ChangelogIterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> identifierFieldIdx) {
+    this.rowIterator = rowIterator;
+    this.changeTypeIndex = changeTypeIndex;
+    this.identifierFieldIdx = identifierFieldIdx;
+  }
+
+  /**
+   * Creates an iterator for records of a changelog table.
+   *
+   * @param rowIterator the iterator of rows from a changelog table
+   * @param changeTypeIndex the index of the change type column
+   * @param identifierFieldIdx the indices of the identifier columns, which determine if rows are
+   *     the same
+   * @return a new {@link ChangelogIterator} instance concatenated with the null-removal iterator
+   */
+  public static Iterator<Row> iterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> identifierFieldIdx) {
+    ChangelogIterator changelogIterator =
+        new ChangelogIterator(rowIterator, changeTypeIndex, identifierFieldIdx);
+    return Iterators.filter(changelogIterator, Objects::nonNull);
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there is an updated cached row, return it directly
+    if (cachedUpdateRecord(cachedRow)) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    }
+
+    Row currentRow = currentRow();
+
+    if (currentRow.getString(changeTypeIndex).equals(DELETE) && rowIterator.hasNext()) {
+      GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+      cachedRow = nextRow;
+
+      if (isUpdateOrCarryoverRecord(currentRow, nextRow)) {
+        if (isCarryoverRecord(currentRow, nextRow)) {
+          // set carry-over rows to null for filtering out later
+          currentRow = null;
+          cachedRow = null;
+        } else {
+          Row[] rows = createUpdateChangelog((GenericRowWithSchema) currentRow, nextRow);
+          currentRow = rows[0];
+          cachedRow = rows[1];
+        }
+      }
+    }
+
+    return currentRow;
+  }
+
+  private Row[] createUpdateChangelog(
+      GenericRowWithSchema currentRow, GenericRowWithSchema nextRow) {
+    GenericInternalRow deletedRow = new GenericInternalRow(currentRow.values());
+    GenericInternalRow insertedRow = new GenericInternalRow(nextRow.values());
+
+    deletedRow.update(changeTypeIndex, UPDATE_BEFORE);
+    insertedRow.update(changeTypeIndex, UPDATE_AFTER);
+
+    return new Row[] {
+      RowFactory.create(deletedRow.values()), RowFactory.create(insertedRow.values())
+    };
+  }
+
+  private boolean isCarryoverRecord(Row currentRow, Row nextRow) {
+    int length = currentRow.length();
```

Comment:
This could probably be optimized a bit. I think the easiest thing to do off the bat is just save row length for the whole iterator rather than checking for every row.

In the future I guess you could do something like 
ColumnsToCompare = [0,1,2,4,5] 
But i'm not sure if that's actually faster. I assume eventually JIT would optimize around the 
```java
if (i == changeTypeIndex) {
  continue;
}
```
But skipping the branch is always nice

Reply:
It is faster. Tested with a simulation of 100 columns comparison with two options
1. Using continue
2. Using a int[] to store the index

The later always takes only 60-70% time of the former, which is surprisingly faster. I didn't expected so much difference. I made the change anyway in this class.
Here is my test code. 
```
    var left = new ArrayList<Integer>();
    var right = new ArrayList<Integer>();
    for(int i = 0; i < 100; i++) {
      left.add(i);
      right.add(i);
    }

    var start = System.currentTimeMillis();
    var match = false;
    for(int i = 0; i < 1000000; i++) {
      for(int j = 0; j < 100; j++) {
        if( j == 50){
          continue;
        }

        match = left.get(j) == right.get(j);
      }
    }
    var end = System.currentTimeMillis();
    System.out.println(String.format("duration: %s, match %s",  end - start, match));

    var indices = new int[99];
    int k = 0;
    for(int i = 0; i < 100; i++) {
      if(i == 50) {
        continue;
      }
      indices[k++] = i;
    }

    start = System.currentTimeMillis();
    for(int i = 0; i < 1000000; i++) {
      for(int j : indices) {
        match = left.get(j) == right.get(j);
      }
    }
    end = System.currentTimeMillis();
    System.out.println(String.format("duration: %s, match: %s",  end - start, match));
```
Here are results:
```
shell:>loop-perf
duration: 119, match true
duration: 92, match: true
shell:>loop-perf
duration: 138, match true
duration: 86, match: true
shell:>loop-perf
duration: 125, match true
duration: 85, match: true
shell:>loop-perf
duration: 125, match true
duration: 85, match: true
shell:>loop-perf
duration: 126, match true
duration: 85, match: true
shell:>loop-perf
duration: 124, match true
duration: 85, match: true
shell:>loop-perf
duration: 125, match true
duration: 85, match: true
shell:>loop-perf
duration: 126, match true
duration: 85, match: true
shell:>loop-perf
duration: 124, match true
duration: 85, match: true
```